scribevulnpolicy/client.go

```diff
@@ -0,0 +1,152 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Contributor:
+// - Zack Mullaly zack@mozilla.com
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+const (
+	clairListVulnsEndptFmt string = "http://127.0.0.1:6060/v1/namespaces/%s/vulnerabilities?limit=99999"
+	clairGetVulnsEndptFmt  string = "http://127.0.0.1:6060/v1/namespaces/%s/vulnerabilities/%s?fixedIn"
+)
+
+// shortVulnResponse contains the response data we expect from a request for a
+// list of vulnerabilities for a namespace.
+type shortVulnResponse struct {
+	Vulns    []shortVuln `json:"Vulnerabilities"`
+	Error    *string     `json:"Error"`
+	NextPage string      `json:"NextPage"`
+}
+
+// longVulnResponse contains response data we expect from a request for more
+// detailed information about a specific vulnerability.
+type longVulnResponse struct {
+	Vuln longVuln `json:"Vulnerability"`
+}
+
+// shortVuln contains only the information we need about vulnerabilities
+// retrieved when we request a list of vulns for a namespace.
+type shortVuln struct {
+	Name string `json:"Name"`
+}
+
+// longVuln contains the information we need about specific vulnerabilities.
+type longVuln struct {
+	Name        string `json:"Name"`
+	Link        string `json:"Link"`
+	Severity    string `json:"Severity"`
+	Description string `json:"Description"`
+	FixedIn     []fix  `json:"FixedIn"`
+}
+
+// fix contains information about what version of a piece of software fixed a
+// particular vulnerability.
+type fix struct {
+	Name    string `json:"Name"`
+	Version string `json:"Version"`
+}
+
+func listVulnsForNamespace(namespace string) ([]shortVuln, error) {
+	respData := shortVulnResponse{}
+	vulns := []shortVuln{}
+
+	// After we get a first response from Clair, we'll have the first
+	// "NextPage" value, which we can use to page through results if
+	// needed.  Once we *don't* get a "NextPage", we know we're done.
+	page := ""
+
+	for {
+		url := fmt.Sprintf(clairListVulnsEndptFmt, namespace)
+		if page != "" {
+			url = fmt.Sprintf("%s&page=%s", url, page)
+		}
+		response, err := http.Get(url)
+		if err != nil {
+			return []shortVuln{}, err
+		}
+
+		decoder := json.NewDecoder(response.Body)
+		defer response.Body.Close()
```

Comment:
we might not want defer in the loop here, since it is executed on function return this could result in a bunch of open requests hanging around (if a number of loops happen)